src/manager/checkpoint.rs

```diff
@@ -170,70 +177,88 @@ async fn manage_subnet((child, parent): (Subnet, Subnet), stop_notify: Arc<Notif
-                            epoch,
-                            account,
-                            &child,
-                            &child_client,
-                            &parent_client,
-                        )
-                        .await?;
+            // check if it has already been successfully committed.
+            // FIXME: We shouldn't just check if OK, but that the error
+            // received is that the checkpoint is not committed yet.
+            if parent_client
+                .ipc_get_checkpoint(&child.id, epoch)
+                .await
+                .is_ok()
+            {
+                log::debug!("checkpoint for epoch {epoch:} already committed. Nothing to do yet!");
+                // Sleep for an appropriate amount of time before checking the chain head again or return
+                // if a stop notification is received.
+                if !wait_next_iteration(&stop_notify).await? {
+                    return Ok(());
+                }
+                continue;
+            }
+
+            // It's a checkpointing epoch and we may have checkpoints to submit.
+            log::info!(
+                "Checkpointing epoch {} for subnet pair ({}, {})",
+                epoch,
+                child.id,
+                parent.id
+            );
+
+            // First, we check which accounts are in the validator set. This is done by reading
+            // the parent's chain head and requesting the state at that tip set.
+            let parent_head = parent_client.chain_head().await?;
+            // A key assumption we make now is that each block has exactly one tip set. We panic
+            // if this is not the case as it violates our assumption.
+            // TODO: update this logic once the assumption changes (i.e., mainnet)
+            assert_eq!(parent_head.cids.len(), 1);
+            let cid_map = parent_head.cids.first().unwrap().clone();
+            let parent_tip_set = Cid::try_from(cid_map)?;
+
+            let subnet_actor_state = parent_client
+                .ipc_read_subnet_actor_state(&child.id, parent_tip_set)
+                .await?;
+
+            let mut validator_set: HashSet<Address, RandomState> = HashSet::new();
+            match subnet_actor_state.validator_set.validators {
+                None => {}
+                Some(validators) => {
+                    for v in validators {
+                        validator_set.insert(Address::from_str(v.addr.deref())?);
                     }
                 }
+            };
+
+            // Now, for each account defined in the `child` subnet that is in the validator set, we
+            // submit a checkpoint on its behalf.
+            assert_eq!(child_head.cids.len(), 1); // Again, check key assumption
+            let child_tip_set = Cid::try_from(child_head.cids.first().unwrap().clone())?;
+            for account in child.accounts.iter() {
+                if validator_set.contains(account) {
```

Comment:
what happens if none of the accounts is in the `validator_set`, we are empty spining? or we assume there will be an account in the validator set?

Reply:
You just do nothing for this round of checkpointing. You are just waiting in the background for when is your turn to submit a new checkpoint